### DISCUSSION TITLE
FGSM for Sklearn MLPClassifier

### DISCUSSION THREAD

**Discussion by mccarthyajb** [Trusted-AI/adversarial-robustness-toolbox#1554] [Q&A]
**Is your feature request related to a problem? Please describe.**
*I am trying to implement FGSM with an MLPClassifier SklearnClassifier  
Unfortunately I get the error message

    art.exceptions.EstimatorError: FastGradientMethod requires an estimator derived from <class  'art.estimators.estimator.BaseEstimator'> and <class 'art.estimators.estimator.LossGradientsMixin'>, the provided classifier is an instance of <class 'art.estimators.classification.scikitlearn.ScikitlearnClassifier'> and is derived from (<class 'art.estimators.classification.classifier.ClassifierMixin'>, <class 'art.estimators.scikitlearn.ScikitlearnEstimator'>).

**Describe the solution you'd like**
I expect FGSM should work with Sklearn MLPClassifier. I wonder if MLPClassifier is currently supported?

**Describe alternatives you've considered**
I have considered building and using using a KerasClassifier instead, and had some success generating FGSM examples.

**Additional context**
    Code Block

    orig_model = MLPClassifier(
            hidden_layer_sizes=(256,128,64,32),
            activation="relu",
           random_state=42,verbose=True,
            max_iter=epochs)

    orig_model.fit(x_train,y_train)
    
    # Step 3: Create the ART classifier

    art_classifier = SklearnClassifier(orig_model)
    #art_classifier.fit(x_train,y_train)

    # Step 6: Generate adversarial test examples

    if(DO_FGSM):
        attack = FastGradientMethod(estimator=art_classifier, eps=0.5)
        print("Generating FGSM")
        x_test_adv = attack.generate(x=x_test)

**Comment by beat-buesser** [comment]
Hi @mccarthyajb Thank you very much for exploring ART! Yes, `MLPClassifier` is not yet supported, the reasons for this are likely that most users of ART would create their MLP models in one of the deep learning frameworks, a second reason is that to support `MLPClassifier` we would have to explicitly  implement its gradient back-propagation function which the deep learning frameworks provide automatically for free. If you would be interested to implement support for `MLPClassifier` and its gradient calculation let me know, we would be happy to provide support.

To run your experiments, I would recommend creating the model in a deep learning framework. You have mentioned `KerasClassifier`, I would recommend its get-started example: https://github.com/Trusted-AI/adversarial-robustness-toolbox/blob/main/examples/get_started_keras.py
